File: 701. 二叉搜索树中的插入操作/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if(root == NULL) return new TreeNode(val);
        traversal(root, val);
        return root;
    }
    void traversal(TreeNode* root, int val){
        if(root == NULL){
            return;
        }
        if(val > root->val){
            if(!root->right){
                root->right = new TreeNode(val);
            }
            traversal(root->right, val);
        }
        if(val < root->val){
            if(!root->left){
                root->left = new TreeNode(val);
            }
            traversal(root->left, val);
        }
        return;
    }
};

File: 701. 二叉搜索树中的插入操作/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traversal(self, root, val):
        if not root:
            return
        if val > root.val:
            if not root.right:
                root.right = TreeNode(val)
            self.traversal(root.right, val)
        if val < root.val:
            if not root.left:
                root.left = TreeNode(val)
            self.traversal(root.left, val)
        return

    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        # 最简单的方式就是不改变树的结构，碰到空节点插入即可
        if not root:
            return TreeNode(val)
        self.traversal(root, val)
        return root